internal/mcp tests don't compile: Config.RateLimit is dead field, newRateLimiter sig mismatch

## Symptom

`go vet ./internal/mcp/` (and any `go test` in that package) fails to build the test binary on `main` (commit d07676d):

\`\`\`
internal/mcp/server_test.go:30:28: too many arguments in call to newRateLimiter
	have (context.Context, Config)
	want (context.Context)
internal/mcp/server_test.go:59:28: too many arguments in call to newRateLimiter
… (and 5 more sites in server_test.go and health_test.go)
\`\`\`

## Root cause

Two API generations were merged without reconciliation:

1. Commit \`31ab4c0\` ("fix: make rate limiter configurable — 0 disables for loopback deployments") added \`Config.RateLimit float64\` and updated tests to pass \`Config{RateLimit: 0}\` / \`Config{RateLimit: 10}\` to a then-Config-aware \`newRateLimiter\`.
2. Commit \`d4cd6bb\` ("fix(server): make rate limiter configurable via env vars, auto-disable on loopback (#389)") replaced the API: \`newRateLimiter(ctx)\` is now single-arg, the live config fields are \`RateLimitRPS\`/\`RateLimitBurst\`/\`RateLimitDisable\`, and \`Config.RateLimit\` is no longer read anywhere.

The test file was never updated, so the package's test binary has not compiled since #389 merged. CI must have a hole that lets this through.

## Fallout

- \`internal/mcp\` test suite cannot run.
- \`Config.RateLimit\` (tools.go:73) is dead code referenced only by the broken tests.
- The two tests \`TestRateLimiter_DisabledWhenConfigIsZero\` and \`TestRateLimiter_RespectsBurstMultiplier\` assert behavior that has no production wiring (the dead field). The Config-aware path is now \`newRateLimiterWithConfig(ctx, cfg.rateLimitRPS(), cfg.rateLimitBurst())\` at server.go:517.

## Fix landing in this PR

- Drop \`, Config{}\` from the 5 trivial call sites that don't read the field.
- Convert \`TestRateLimiter_RespectsBurstMultiplier\` to call \`newRateLimiterWithConfig(ctx, 10, 40)\` directly.
- Skip \`TestRateLimiter_DisabledWhenConfigIsZero\` with a reference to this issue — the "disable via Config" feature is not currently wired (only \`RateLimitDisable\` at the HTTP middleware layer is, and it's not exercised through the rate-limiter constructor).

## Follow-up (not in this PR)

- Decide: delete \`Config.RateLimit\` (the field) as dead code, or wire it up as the canonical disable knob and deprecate \`RateLimitDisable\`. Currently we have three overlapping rate-limit knobs.
- Add a CI check that \`go vet ./...\` passes — this would have caught the regression at #389 merge time.

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -39,3 +39,7 @@ memory_aggregate(by="failure_class")
 Valid `failure_class` values: `vocabulary_mismatch`, `aggregation_failure`, `stale_ranking`, `missing_content`, `scope_mismatch`, `other`.
 
 Use `memory_aggregate(by="failure_class")` periodically to see where recall is failing most. This data feeds retrieval quality benchmarking.
+
+## MCP Read-Only Hint Annotations
+
+Read-side tools (recall, fetch, query, list, status, history, timeline, projects, episode/audit listings, constraint checks, diagnose) carry `ReadOnlyHint: true` on their MCP annotation. The canonical set lives in `readOnlyToolNames()` in `internal/mcp/server.go` — add new read-only tools there, not by editing `registerTools` directly. The annotation is what lets Claude Code's plan mode invoke these tools without prompting; without it, calls are silently rejected client-side. `TestReadOnlyToolAnnotations` (`internal/mcp/readonly_hints_test.go`) is the regression guard.

--- a/README.md
+++ b/README.md
@@ -65,6 +65,24 @@ The server starts on port 8788. If you prefer to author `.env` by hand rather th
 
 Run `/mcp` in Claude Code after setup to connect. All 38 core tools are available immediately. Five optional AI-enhanced tools (`memory_ask`, `memory_reason`, `memory_explore`, `memory_query_document`, `memory_diagnose`) activate when `ANTHROPIC_API_KEY` is set.
 
+### Claude Code permissions
+
+Engram's read-side tools (`memory_recall`, `memory_fetch`, `memory_query`, `memory_list`, `memory_status`, `memory_history`, `memory_timeline`, `memory_projects`, audit/episode listings, constraint checks, `memory_diagnose`) ship with the MCP `ReadOnlyHint: true` annotation. Claude Code's plan mode treats them as safe to invoke without a permission prompt — recall just works, even from inside plan mode.
+
+Mutating tools (`memory_store`, `memory_correct`, `memory_forget`, `memory_consolidate`, `memory_delete_project`, etc.) intentionally prompt for permission on first use. To skip prompts on the read-only tools in normal mode, add the recommended snippet to `~/.claude/settings.json`:
+
+```json
+{
+  "permissions": {
+    "allow": ["mcp__engram__*"]
+  }
+}
+```
+
+The engram server logs a paste-ready `permissions.allow` snippet at startup containing the exact list of read-only tool names — copy from the log if you want a narrower allowlist than the wildcard.
+
+If a recall tool ever vanishes silently (no result, no prompt, no error), the call was almost certainly blocked client-side. Check that you're running an engram build that includes the `ReadOnlyHint` annotation (engram-go ≥ this PR) and that no `permissions.deny` rule shadows the engram tool name.
+
 ---
 
 ## Architecture

--- a/cmd/engram/main.go
+++ b/cmd/engram/main.go
@@ -387,6 +387,8 @@ func run() error {
 		slog.Warn("session rehydration failed — clients will need to reconnect", "err", err)
 	}
 
+	logRecommendedClientPermissions(srv)
+
 	slog.Info("engram ready", "host", *host, "port", *port,
 		"embed_model", *embedModel, "summarize_model", sumModel)
 	return srv.Start(ctx, *host, *port, apiKey, *baseURL)

--- a/cmd/engram/permissions_banner.go
+++ b/cmd/engram/permissions_banner.go
@@ -1,0 +1,45 @@
+package main
+
+import (
+	"encoding/json"
+	"log/slog"
+	"sort"
+
+	internalmcp "github.com/petersimmons1972/engram/internal/mcp"
+)
+
+// logRecommendedClientPermissions emits a one-time INFO log naming the read-only
+// MCP tools that benefit from being on the Claude Code permissions allowlist.
+// Operators can copy the JSON snippet directly out of the log into
+// ~/.claude/settings.json.
+//
+// Why this exists: even with ReadOnlyHint=true on the tool annotation, an
+// over-cautious deny rule or non-default permission policy can still block
+// these calls. Showing the canonical allowlist at startup is the cheapest way
+// to keep installers from hitting silent rejection loops.
+func logRecommendedClientPermissions(srv *internalmcp.Server) {
+	annotations := srv.RegisteredToolAnnotations()
+	names := make([]string, 0, len(annotations))
+	for name, ann := range annotations {
+		if ann.ReadOnlyHint != nil && *ann.ReadOnlyHint {
+			names = append(names, "mcp__engram__"+name)
+		}
+	}
+	sort.Strings(names)
+
+	snippet := map[string]any{
+		"permissions": map[string]any{
+			"allow": names,
+		},
+	}
+	body, err := json.MarshalIndent(snippet, "", "  ")
+	if err != nil {
+		slog.Warn("could not render recommended permissions snippet", "err", err)
+		return
+	}
+
+	slog.Info("recommended Claude Code permissions for engram read-only tools — paste into ~/.claude/settings.json under permissions.allow",
+		"count", len(names),
+		"snippet", string(body),
+	)
+}

--- a/internal/mcp/health_test.go
+++ b/internal/mcp/health_test.go
@@ -167,7 +167,7 @@ func TestApplyMiddleware_SetsRequestID(t *testing.T) {
 	defer cancel()
 
 	s := &Server{}
-	rl := newRateLimiter(ctx, Config{})
+	rl := newRateLimiter(ctx)
 	apiKey := "test-key"
 
 	var capturedID string

--- a/internal/mcp/readonly_hints_test.go
+++ b/internal/mcp/readonly_hints_test.go
@@ -1,0 +1,75 @@
+package mcp
+
+import (
+	"context"
+	"testing"
+
+	mcpgo "github.com/mark3labs/mcp-go/mcp"
+	"github.com/mark3labs/mcp-go/server"
+)
+
+// TestReadOnlyToolAnnotations verifies that every tool listed in
+// readOnlyToolNames() carries ReadOnlyHint=true on its MCP annotation, and
+// that representative mutating tools carry ReadOnlyHint=false.
+//
+// Why this exists: clients (notably Claude Code's plan mode) gate tool calls
+// on the ReadOnlyHint annotation. If a recall tool ships without the hint,
+// the call is silently rejected client-side with no permission prompt — the
+// failure mode that motivated this work. This test prevents the regression
+// recurring when new tools are added.
+func TestReadOnlyToolAnnotations(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	srv := &Server{
+		cfg:             Config{},
+		uploads:         make(map[string]*uploadSession),
+		toolAnnotations: make(map[string]mcpgo.ToolAnnotation),
+	}
+	srv.mcp = server.NewMCPServer("engram-test", "0.0.0",
+		server.WithToolCapabilities(true),
+		server.WithHooks(&server.Hooks{}),
+	)
+	srv.registerTools()
+
+	annotations := srv.RegisteredToolAnnotations()
+	expectedReadOnly := readOnlyToolNames()
+
+	// Every read-only tool must have been registered AND carry ReadOnlyHint=true.
+	for name := range expectedReadOnly {
+		ann, ok := annotations[name]
+		if !ok {
+			t.Errorf("tool %q is in readOnlyToolNames() but was not registered (typo or removed?)", name)
+			continue
+		}
+		if ann.ReadOnlyHint == nil || !*ann.ReadOnlyHint {
+			t.Errorf("tool %q: expected ReadOnlyHint=true, got %v", name, ann.ReadOnlyHint)
+		}
+	}
+
+	// Representative mutating tools must NOT have ReadOnlyHint=true. Asserting on
+	// a small representative set rather than the inverse of readOnlyToolNames so
+	// the test stays useful even if new mutating tools are added without updating
+	// the read-only set.
+	mutators := []string{
+		"memory_store",
+		"memory_correct",
+		"memory_forget",
+		"memory_delete_project",
+		"memory_consolidate",
+		"memory_feedback",
+	}
+	for _, name := range mutators {
+		ann, ok := annotations[name]
+		if !ok {
+			t.Errorf("expected mutator tool %q to be registered", name)
+			continue
+		}
+		if ann.ReadOnlyHint != nil && *ann.ReadOnlyHint {
+			t.Errorf("tool %q is a mutator but carries ReadOnlyHint=true — it will bypass plan-mode gating", name)
+		}
+	}
+
+	// Unused ctx cancel guard (linter quiet).
+	_ = ctx
+}

--- a/internal/mcp/server.go
+++ b/internal/mcp/server.go
@@ -154,6 +154,61 @@ type Server struct {
 	sessionFingerprints sync.Map // sessionID -> []byte HMAC fingerprint (#245)
 	sessionEpisodes     sync.Map // sessionID -> episodeID string for auto-episode sessions (#356)
 	trustProxy          bool     // honour X-Forwarded-For / X-Real-IP (#255)
+
+	// toolAnnotations records the MCP ToolAnnotation set on each registered tool.
+	// Populated as a side-effect of registerToolWithTimeout. Read by the startup
+	// banner (recommended permissions snippet) and by tests to verify that
+	// read-only tools carry ReadOnlyHint=true so client-side gating (e.g. Claude
+	// Code's plan mode) does not block them.
+	toolAnnotations map[string]mcpgo.ToolAnnotation
+}
+
+// readOnlyToolNames returns the canonical set of MCP tool names that perform
+// no mutations to memory state. These tools get ReadOnlyHint=true on their
+// MCP annotation so clients (notably Claude Code's plan mode) treat them as
+// safe to invoke without a permission prompt.
+//
+// Single source of truth — used by both registerTools (to set the annotation)
+// and the startup banner (to print a recommended permissions.allow snippet).
+func readOnlyToolNames() map[string]bool {
+	return map[string]bool{
+		"memory_recall":             true,
+		"memory_query":              true,
+		"memory_fetch":              true,
+		"memory_list":               true,
+		"memory_history":            true,
+		"memory_timeline":           true,
+		"memory_status":             true,
+		"memory_projects":           true,
+		"memory_aggregate":          true,
+		"memory_diagnose":           true,
+		"memory_episode_list":       true,
+		"memory_episode_recall":     true,
+		"memory_models":             true,
+		"memory_embedding_eval":     true,
+		"memory_export_all":         true,
+		"memory_audit_list_queries": true,
+		"memory_audit_compare":      true,
+		"memory_audit_run":          true,
+		"memory_weight_history":     true,
+		"memory_ingest_status":      true,
+		"memory_expand":             true,
+		"memory_verify":             true,
+		"get_constraints":           true,
+		"check_constraints":         true,
+		"verify_before_acting":      true,
+	}
+}
+
+// RegisteredToolAnnotations returns a copy of the ToolAnnotation that was
+// applied to each registered tool. Intended for tests and for the startup
+// banner. The returned map is safe to mutate.
+func (s *Server) RegisteredToolAnnotations() map[string]mcpgo.ToolAnnotation {
+	out := make(map[string]mcpgo.ToolAnnotation, len(s.toolAnnotations))
+	for k, v := range s.toolAnnotations {
+		out[k] = v
+	}
+	return out
 }
 
 // NewServer constructs a Server with all MCP tools registered.
@@ -163,7 +218,13 @@ func NewServer(pool *EnginePool, cfg Config) *Server {
 		trustProxy = true
 		slog.Warn("ENGRAM_TRUST_PROXY_HEADERS is enabled — ensure a trusted reverse proxy terminates all inbound connections; direct clients can spoof X-Forwarded-For to bypass rate limiting")
 	}
-	s := &Server{pool: pool, cfg: cfg, uploads: make(map[string]*uploadSession), trustProxy: trustProxy}
+	s := &Server{
+		pool:            pool,
+		cfg:             cfg,
+		uploads:         make(map[string]*uploadSession),
+		trustProxy:      trustProxy,
+		toolAnnotations: make(map[string]mcpgo.ToolAnnotation),
+	}
 	mcpServer := server.NewMCPServer("engram", "1.0.0",
 		server.WithToolCapabilities(true),
 		server.WithHooks(&server.Hooks{}),
@@ -756,15 +817,28 @@ func withWarnLog(name string, h toolHandler) toolHandler {
 const defaultToolTimeout = 15 * time.Second
 
 // registerToolWithTimeout adds a tool to the MCP server with a per-call deadline
-// and Prometheus instrumentation. timeout=0 uses defaultToolTimeout.
-func (s *Server) registerToolWithTimeout(name, desc string, h toolHandler, timeout time.Duration) {
+// and Prometheus instrumentation. timeout=0 uses defaultToolTimeout. readOnly
+// sets the MCP ReadOnlyHint annotation: clients (notably Claude Code's plan
+// mode) use that hint to decide whether to invoke a tool without prompting.
+func (s *Server) registerToolWithTimeout(name, desc string, h toolHandler, timeout time.Duration, readOnly bool) {
 	if timeout == 0 {
 		timeout = defaultToolTimeout
 	}
 	pool, cfg := s.pool, s.cfg
 	toolName := name
 	toolTimeout := timeout
-	s.mcp.AddTool(mcpgo.NewTool(name, mcpgo.WithDescription(desc)),
+	roCopy := readOnly
+	annotation := mcpgo.ToolAnnotation{
+		Title:        name,
+		ReadOnlyHint: &roCopy,
+	}
+	if s.toolAnnotations == nil {
+		s.toolAnnotations = make(map[string]mcpgo.ToolAnnotation)
+	}
+	s.toolAnnotations[name] = annotation
+	s.mcp.AddTool(mcpgo.NewTool(name,
+		mcpgo.WithDescription(desc),
+		mcpgo.WithToolAnnotation(annotation)),
 		func(ctx context.Context, req mcpgo.CallToolRequest) (*mcpgo.CallToolResult, error) {
 			ctx, cancel := context.WithTimeout(ctx, toolTimeout)
 			defer cancel()
@@ -781,7 +855,10 @@ func (s *Server) registerToolWithTimeout(name, desc string, h toolHandler, timeo
 					Content: []mcpgo.Content{
 						mcpgo.TextContent{
 							Type: "text",
-							Text: toolName + " timed out after " + toolTimeout.String() + " — Ollama may be slow or unavailable",
+							Text: toolName + " timed out after " + toolTimeout.String() +
+								" — backend (Ollama/LiteLLM) may be slow or unavailable. " +
+								"If you did not see a permission prompt either, the call may have been blocked client-side " +
+								"(e.g. Claude Code plan mode or a deny rule) before reaching engram; check ~/.claude/settings.json permissions and current mode.",
 						},
 					},
 				}, nil
@@ -795,9 +872,10 @@ func (s *Server) registerToolWithTimeout(name, desc string, h toolHandler, timeo
 		})
 }
 
-// registerTool adds a tool with the default dispatch timeout.
+// registerTool adds a tool with the default dispatch timeout. The readOnly hint
+// is sourced from readOnlyToolNames() — single source of truth.
 func (s *Server) registerTool(name, desc string, h toolHandler) {
-	s.registerToolWithTimeout(name, desc, h, 0)
+	s.registerToolWithTimeout(name, desc, h, 0, readOnlyToolNames()[name])
 }
 
 func (s *Server) registerTools() {

--- a/internal/mcp/server_test.go
+++ b/internal/mcp/server_test.go
@@ -27,7 +27,7 @@ func TestSetupTokenBudgetIsolatedFromNormalBudget(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	rl := newRateLimiter(ctx, Config{})
+	rl := newRateLimiter(ctx)
 	ip := "10.0.0.5"
 
 	// Exhaust the normal budget by calling allow() many times.
@@ -56,7 +56,7 @@ func TestSetupTokenBudgetDoesNotConsumeNormalBudget(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	rl := newRateLimiter(ctx, Config{})
+	rl := newRateLimiter(ctx)
 	ip := "10.0.0.6"
 
 	// Exhaust setup-token budget.
@@ -185,7 +185,7 @@ func TestRateLimiter_AllowSetupToken(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	rl := newRateLimiter(ctx, Config{})
+	rl := newRateLimiter(ctx)
 	ip := "192.168.1.1"
 
 	// First 3 calls must succeed — burst of 3 tokens.
@@ -207,7 +207,7 @@ func TestRateLimiter_SetupTokenIndependentPerIP(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	rl := newRateLimiter(ctx, Config{})
+	rl := newRateLimiter(ctx)
 
 	// Exhaust setup-token budget for ip1.
 	for i := 0; i < 3; i++ {
@@ -220,32 +220,24 @@ func TestRateLimiter_SetupTokenIndependentPerIP(t *testing.T) {
 	}
 }
 
-// TestRateLimiter_DisabledWhenConfigIsZero verifies that when Config.RateLimit=0,
-// the rate limiter allows unlimited requests (always returns true).
+// TestRateLimiter_DisabledWhenConfigIsZero is skipped pending issue #422.
+// The "Config.RateLimit=0 means unlimited" feature it asserts is not currently
+// wired through the rate-limiter constructor — Config.RateLimit is a dead field.
 func TestRateLimiter_DisabledWhenConfigIsZero(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	// Config with RateLimit=0 means unlimited.
-	rl := newRateLimiter(ctx, Config{RateLimit: 0})
-	ip := "10.0.0.7"
-
-	// Make many calls — all should succeed (no rate limit enforced).
-	for i := 0; i < 1000; i++ {
-		if !rl.allow(ip) {
-			t.Fatalf("call %d: expected allow with RateLimit=0, got reject", i)
-		}
-	}
+	t.Skip("Config.RateLimit field is dead — see issue #422")
 }
 
-// TestRateLimiter_RespectsBurstMultiplier verifies that when RateLimit is set,
-// the burst is calculated as 4× the rate, and requests respect that burst budget.
+// TestRateLimiter_RespectsBurstMultiplier verifies that when burst is set,
+// requests respect that burst budget. Uses newRateLimiterWithConfig directly
+// since Config.RateLimit is dead (issue #422); behavior under test is the
+// burst-capacity logic of the per-IP rate.Limiter, which is unchanged.
 func TestRateLimiter_RespectsBurstMultiplier(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	// Set RateLimit to 10 req/s, which gives burst of 40 (4×).
-	rl := newRateLimiter(ctx, Config{RateLimit: 10})
+	// 10 req/s sustained, burst of 40 — same shape the dead Config.RateLimit
+	// path was meant to produce (burst = 4 × RateLimit).
+	rl := newRateLimiterWithConfig(ctx, 10, 40)
 	ip := "10.0.0.8"
 
 	// First 40 calls should succeed (burst).
@@ -580,7 +572,7 @@ func TestDispatchTimeout_RegisterToolAppliesDeadline(t *testing.T) {
 	}
 
 	// Register with a short timeout so the test completes quickly.
-	srv.registerToolWithTimeout("block_test", "blocks until ctx cancelled", blockingHandler, 200*time.Millisecond)
+	srv.registerToolWithTimeout("block_test", "blocks until ctx cancelled", blockingHandler, 200*time.Millisecond, false)
 
 	c, err := mcpclient.NewInProcessClient(srv.mcp)
 	if err != nil {

